catalyst_voices/apps/voices/lib/pages/proposal/widget/proposal_comment_builder.dart

```diff
@@ -0,0 +1,178 @@
+  final VoidCallback? onCancelTap;
+  final bool showCancel;
+
+  const _Actions({
+    this.onSubmitTap,
+    this.onCancelTap,
+    required this.showCancel,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    return Row(
+      spacing: 8,
+      mainAxisAlignment: MainAxisAlignment.end,
+      children: [
+        if (showCancel)
+          VoicesTextButton(
+            onTap: onCancelTap,
+            child: Text(context.l10n.cancelCommit),
+          ),
+        VoicesFilledButton(
+          onTap: onSubmitTap,
+          child: Text(context.l10n.submitCommit),
+        ),
+      ],
+    );
+  }
+}
+
+class _ProposalCommentBuilderState extends State<ProposalCommentBuilder> {
+  final _formKey = GlobalKey<FormState>();
+  final _focusNode = FocusScopeNode();
+
+  late DocumentBuilder _builder;
+  late Document _comment;
+
+  @override
+  Widget build(BuildContext context) {
+    return DocumentBuilderTheme(
+      data: const DocumentBuilderThemeData(
+        shouldDebounceChange: false,
+      ),
+      child: Row(
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: [
+          const _SessionAccountAvatar(),
+          const SizedBox(width: 16),
+          Expanded(
+            child: Form(
+              key: _formKey,
+              autovalidateMode: AutovalidateMode.onUserInteraction,
+              child: FocusScope(
+                node: _focusNode,
+                child: Column(
+                  mainAxisSize: MainAxisSize.min,
+                  children: [
+                    for (final property in _comment.properties)
+                      DocumentPropertyBuilder(
+                        key: ValueKey(property.nodeId),
+                        property: property,
+                        isEditMode: true,
+                        onChanged: _handlePropertyChanges,
+                      ),
+                    const SizedBox(height: 4),
+                    _Actions(
+                      onSubmitTap: _submit,
+                      onCancelTap: widget.onCancelTap,
+                      showCancel: widget.showCancel,
+                    ),
+                  ],
+                ),
+              ),
+            ),
+          ),
+        ],
+      ),
+    );
+  }
+
+  @override
+  void didUpdateWidget(covariant ProposalCommentBuilder oldWidget) {
+    super.didUpdateWidget(oldWidget);
+
+    if (widget.schema != oldWidget.schema) {
+      _builder = DocumentBuilder.fromSchema(schema: widget.schema);
+      _comment = _builder.build();
+    }
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    _focusNode.dispose();
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    _builder = DocumentBuilder.fromSchema(schema: widget.schema);
+    _comment = _builder.build();
+  }
+
+  void _handlePropertyChanges(List<DocumentChange> changes) {
+    setState(() {
+      for (final change in changes) {
+        _builder.addChange(change);
+      }
```

Comment:
It should be possible to apply them in batch:
```suggestion
       _builder.addChanges(changes);
```